File: c++/src/agent.cc
Agent::Agent() {
	x = 0;
	y = 0;
	width = 10; // Todo: make these realistic default values
	height = 10;

	active = false;
}

Agent::Agent(int newx, int newy, int newwidth, int newheight) {
	x = newx;
	y = newy;

	width = newwidth;
	height = newheight;

	active = false;
}

int Agent::fitness() {
	return Search::getInstance().fitness(x, y, width, height);
}

void Agent::update_state() {
	int percentage = fitness();

	//if(percentage <= Agents::getInstance().get_best_percentage()) {
	if(percentage < 15) {
		Agents::getInstance().new_percentage(percentage);
		active = true;

		if(rand() % 10 == 0) {
			active = false;
		}
	}
	else {
		active = false;
	}
}

void Agent::update() {
	if(active == true) {
		if(rand() % 10 == 0) {
			move_vicinity();
		}
	}
	else {
		Agent random_agent = Agents::getInstance().random_agent();

		if(random_agent.is_active()) {
			x = random_agent.get_x();
			y = random_agent.get_y();
		}
		else {
			if(rand() % 2 == 0) {
				randomize();
			}
			else {
				move_vicinity();
			}
		}
	}
}

void Agent::randomize() {
	// x and y have to be within the boundaries of the haystack image set for this node
	x = Search::getInstance().node_x + (rand() % Search::getInstance().node_width);
	y = Search::getInstance().node_y + (rand() % Search::getInstance().node_height);
}

void Agent::move_vicinity() {
	x  += (rand() % 21) - 10;
	y += (rand() % 21) - 10;
}

void Agent::draw() {
	int color = SDL_MapRGB(SDL::getInstance().screen->format, 255, 0, 0);
	if(active == true) {
		color = SDL_MapRGB(SDL::getInstance().screen->format, 0, 255, 0);
	}

	int draw_x, draw_y;
	draw_x = BORDER_X + (x - Search::getInstance().node_x) * PIXEL_SIZE;
	draw_y = BORDER_Y + (y - Search::getInstance().node_y) * PIXEL_SIZE;

	// Draw in the fitness percentage
	SDL_Color textColor = { 255, 255, 255};
	SDL_Surface *text = TTF_RenderText_Solid(SDL::getInstance().font, intToString(fitness()), textColor);

	SDL::getInstance().applySurface(text, draw_x, draw_y);

	SDL::getInstance().drawRect(draw_x, draw_y, width * PIXEL_SIZE, height * PIXEL_SIZE, 1, color);
}

bool Agent::is_active() {
	return active;
}

int Agent::get_x() {
	return x;
}

int Agent::get_y() {
	return y;
}

File: ruby/agent.rb
class Agent
	attr_accessor :x, :y

	def initialize(x, y, width, height)
		@x = x
		@y = y
		@width = width
		@height = height

		@active = false

		self
	end

	def active?
		return @active
	end
	
	def distance()
		return Search::instance.distance(@x, @y, @width, @height)
	end

	def distance_percent()
		return (100 * distance()) / Search::instance.max_distance
	end

	def draw()
		border_color = Graphics::instance.screen.mapRGB(255, 0, 0)
		if @active == true
			border_color = Graphics::instance.screen.mapRGB(0, 255, 0)
		end

		border_width = AGENT_BORDER_WIDTH_INACTIVE
		border_width = AGENT_BORDER_WIDTH_ACTIVE if active?

		border_width.times do |i|
			Graphics::instance.screen.draw_rect(i + BORDER_X + @x * PIXEL_SIZE, i + BORDER_Y + @y * PIXEL_SIZE, @width * PIXEL_SIZE - 1, @height * PIXEL_SIZE - 1, border_color)
		end

		#Graphics::instance.font.drawBlendedUTF8(Graphics::instance.screen, distance_percent().to_s, BORDER_X + @x * PIXEL_SIZE, BORDER_Y + @y * PIXEL_SIZE, 150, 150, 150)
	end

	def randomize
		@x = (rand() * Search::instance.haystack.columns).to_i
		@y = (rand() * Search::instance.haystack.rows).to_i
		self
	end

	def update_state
		if distance_percent <= Agents::instance.best_percentage
			Agents::instance.new_percentage(distance_percent)

			@active = true

			# Small chance the agent will randomly switch to inactive
			if rand(10) == 1
				@active = false
			end
		else
			@active = false
		end
	end

	def move_vicinity
		# The agent can move either -1, 0, or 1 pixels in any direction
		dx = rand(3) - 1
		dy = rand(3) - 1

		@x += dx
		@y += dy
	end

	def update
		# for now, just randomly change position
		if @active
			# Have a random chance that the agent will move to a slightly different location in the immediate vicinity
			if rand(10) == 1
				#move_vicinity
				#update_state
			end
		else
			random_agent = Agents::instance.random_agent
			
			if random_agent.active?
				@x = random_agent.x
				@y = random_agent.y
			else
				if rand(3) == 1
					randomize
				else
					move_vicinity
				end
			end
		end
	end
end
